hub/core/tests/test_storage_provider.py

```diff
@@ -34,18 +40,197 @@ def test_provider(provider=MappedProvider()):
     with pytest.raises(KeyError):
         provider[FILE_1]
 
+    provider.flush()
+
 
 def test_memory_provider():
-    test_provider(MemoryProvider("abc/def"))
+    check_storage_provider(memory_provider)
 
 
 def test_local_provider():
-    test_provider(LocalProvider("./hub_storage_local_test"))
+    check_storage_provider(local_provider)
 
 
 @pytest.mark.skipif(not s3_creds_exist(), reason="requires s3 credentials")
 def test_s3_provider():
-    test_provider(S3Provider("snark-test/hub_storage_s3_test"))
+    check_storage_provider(s3_provider)
+
+
+def test_lru_mem_local(benchmark):
+    lru = get_cache_chain([memory_provider, local_provider], [32 * MB])
+    check_storage_provider(lru)
+
+
+@pytest.mark.skipif(not s3_creds_exist(), reason="requires s3 credentials")
+def test_lru_mem_s3(benchmark):
+    lru = get_cache_chain([memory_provider, s3_provider], [32 * MB])
+    check_storage_provider(lru)
+
+
+@pytest.mark.skipif(not s3_creds_exist(), reason="requires s3 credentials")
+def test_lru_local_s3(benchmark):
+    lru = get_cache_chain([local_provider, s3_provider], [160 * MB])
+    check_storage_provider(lru)
+
+
+@pytest.mark.skipif(not s3_creds_exist(), reason="requires s3 credentials")
+def test_lru_mem_local_s3(benchmark):
+    lru = get_cache_chain(
+        [memory_provider, local_provider, s3_provider],
+        [32 * MB, 160 * MB],
+    )
+    check_storage_provider(lru)
+
+
+def write_to_files(provider):
+    chunk = b"0123456789123456" * MB
+    for i in range(20):
+        provider[f"file_{i}"] = chunk
+    provider.flush()
+
+
+def read_from_files(provider):
+    for i in range(20):
+        provider[f"file_{i}"]
+
+
+def delete_files(provider):
+    for i in range(20):
```

Comment:
We should factor out the `20` so it's not a magic constant repeated in multiple places. It also can't hurt to change this to `for k in provider` as well